### PR TITLE
Add fixit for '-> throws' to 'throws ->'

### PR DESCRIPTION
**Pull request by harlanhaskins**
#### What's in this pull request?

Update function signature parsing to handle `throws` between the arrow and the type.

This will fix:

``` swift
func foo() -> throws Int
```

to

``` swift
func foo() throws -> Int
```

and remove the bogus parse errors that are produced for this now.
#### Resolved bug number: ([SR-1940](https://bugs.swift.org/browse/SR-1940))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
@gribozavr Good to merge?

File: lib/Parse/ParsePattern.cpp [hunk 1]
```diff
@@ -648,6 +648,26 @@ Parser::parseFunctionSignature(Identifier SimpleName,
...
+    if (Tok.is(tok::kw_throws)) {
+      throwsLoc = consumeToken();
+    } else if (Tok.is(tok::kw_rethrows)) {
+      throwsLoc = consumeToken();
+      rethrows = true;
+    }
+
+    if (throwsLoc.isValid()) {
+      auto diag = rethrows ? diag::rethrows_after_function_result
+                           : diag::throws_after_function_result;
```
Comment (gribozavr) [inline lib/Parse/ParsePattern.cpp]: The names of these diagnostics are a little misleading now. `throws_in_incorrect_place`?

File: lib/Parse/ParsePattern.cpp [hunk 2]
```diff
@@ -648,6 +648,26 @@ Parser::parseFunctionSignature(Identifier SimpleName,
...
+      return None;
+
+    if (Tok.is(tok::kw_throws)) {
+      throwsLoc = consumeToken();
+    } else if (Tok.is(tok::kw_rethrows)) {
+      throwsLoc = consumeToken();
+      rethrows = true;
+    }
+
+    if (throwsLoc.isValid()) {
```
Comment (gribozavr) [inline lib/Parse/ParsePattern.cpp]: You can also do `if (!throwsLoc.isValid()) return None;` and de-indent.

File: lib/Parse/ParsePattern.cpp [hunk 3]
```diff
@@ -672,26 +703,16 @@ Parser::parseFunctionSignature(Identifier SimpleName,
...
-                           : diag::throws_after_function_result;
-      SourceLoc typeEndLoc = Lexer::getLocForEndOfToken(SourceMgr,
-                                                        retType->getEndLoc());
-      SourceLoc throwsEndLoc = Lexer::getLocForEndOfToken(SourceMgr, throwsLoc);
-      diagnose(Tok, diag)
-        .fixItInsert(arrowLoc, rethrows ? "rethrows " : "throws ")
-        .fixItRemoveChars(typeEndLoc, throwsEndLoc);
-    }
+  auto diagOpt = diagnoseInvalidThrows();
+  if (diagOpt) {
```
Comment (gribozavr) [inline lib/Parse/ParsePattern.cpp]: `if (auto diagOpt = diagnoseInvalidThrows())`?  Won't unwrap the optional, but at least would reduce the scope of the variable.

**Comment by gribozavr** [comment]
LGTM modulo nitpicks!

File: lib/Parse/ParsePattern.cpp (hunk 1 above)
Comment (harlanhaskins) [inline lib/Parse/ParsePattern.cpp]: Agreed!

File: lib/Parse/ParsePattern.cpp (hunk 3 above)
Comment (harlanhaskins) [inline lib/Parse/ParsePattern.cpp]: Good call. Plus, I wouldn't need to call `.flush()`

File: test/Parse/errors.swift [hunk 4]
```diff
@@ -110,10 +110,18 @@ func postThrows() -> Int throws { // expected-error{{'throws' may only occur bef
   return 5
 }
 
+func postThrows2() -> throws Int { // expected-error{{'throws' may only occur before '->'}}{{20-22=throws}}{{23-29=->}}
+  return 5
```
Comment (jrose-apple) [inline test/Parse/errors.swift]: Can you test that we've recovered as a throwing function as well, by calling something else that throws and not getting a warning?

**Comment by harlanhaskins** [comment]
@gribozavr @jrose-apple Fixed up based on your suggestions

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Pull request merged**